### PR TITLE
blog: TPI - Release post

### PR DESCRIPTION
**Pull request by jendefig**
This is Maria's first post to go out on 2022 April 27, 06:00 PDT

- [x] Discuss link needs to be changed, waiting until closer
- [x] please check the code as I copied it over from Google docs and it was incomplete in there, copied the code from Tutorial post one.
- [x] Image has been tinified
- [x] Needs language check - I made revisions according to discussions in Google Doc.

### CODE REVIEW

File: content/blog/2022-04-18-terraform-provider.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+     machine   = "l+k80"
+     directory = "."
+     script = <<-END
+       #!/bin/bash
+       sudo apt update
+       sudo apt install -y python3-pip
+       pip3 install -r requirements.txt
+       python3 src/train.py
+     END
+ }
```
Comment (0x2b3bfa0) [inline content/blog/2022-04-18-terraform-provider.md:75]: Updated as per the [Getting Started](https://registry.terraform.io/providers/iterative/iterative/latest/docs/guides/getting-started) guide.

```suggestion
terraform {
  required_providers {
    iterative = {
      source = "iterative/iterative"
    }
  }
}

provider "iterative" {}

resource "iterative_task" "example" {
  cloud   = "aws"
  region  = "us-east-2"
  machine = "l+k80"
  
  script = <<-END
    #!/bin/bash
    sudo apt update
    sudo apt install -y python3-pip
    pip3 install -r requirements.txt
    python3 src/train.py
  END
  
  storage {
    workdir = "."
    output  = "results"
  }
}
```

🔔 @casperdcl may want to “casperify” the code format.

**Review by 0x2b3bfa0 — COMMENTED** [review]
API has changed since this post was written

**Comment by jorgeorpinel** [comment]
> This is Maria's first post to go out on April 18, 2022.

So can we close https://github.com/iterative/dvc.org/pull/3087 @jendefig ? p.s. this PR is missing [content/authors/maria_khalusova.md](https://github.com/iterative/dvc.org/pull/3087/files#diff-63dc95889e0bd09c2982273c30044077178870e684b152cefbf8874e202a8bc8), I think.

**Comment by jendefig** [comment]
> > This is Maria's first post
> 
> So can we close #3087 @jendefig ? p.s. this PR is missing [content/authors/maria_khalusova.md](https://github.com/iterative/dvc.org/pull/3087/files#diff-63dc95889e0bd09c2982273c30044077178870e684b152cefbf8874e202a8bc8), I think.

No.  3087 is legit as it is a tutorial, as well as the other one with Docker.  I've pulled 3087 down so that I could actually view it because it wasn't rendering in Heroku because of the author stuff.  I will try to get dates corrected and send back up today 🤞 .  3087 will be April 26th- ish. 

This one is the release post that will go out on April 18th with the release.  It previously only lived in a Google Doc.

> > This is Maria's first post
> 
> So can we close #3087 @jendefig ? p.s. this PR is missing [content/authors/maria_khalusova.md](https://github.com/iterative/dvc.org/pull/3087/files#diff-63dc95889e0bd09c2982273c30044077178870e684b152cefbf8874e202a8bc8), I think.

No. 3087 is a tutorial.  This is the release post.  I've pulled that one down so that I could actually look at it.  I don't know why this doesn't have the content/authors.... It does in my computer.  I will try to get this fixed today and get 3087 up with correct date (April 26th.)

**Comment by jorgeorpinel** [comment]
Ah looks like the author is [already merged into master](https://github.com/iterative/dvc.org/blob/master/content/authors/maria_khalusova.md) so that's OK then, no need to try including that file in this PR.

> In fact that file is causing a conflict on #3087

**Comment by jendefig** [comment]
> Ah looks like the author is [already merged into master](https://github.com/iterative/dvc.org/blob/master/content/authors/maria_khalusova.md) so that's OK then, no need to try including that file in this PR.
> 
> > In fact that file is causing a conflict on #3087

Ya.  I totally give up here.  Still can't see tutorial post.

**Comment by jorgeorpinel** [comment]
You can see the post on the review app in https://dvc-org-tpi-blog-83fa8qsgo4nms.herokuapp.com/blog/terraform-provider

**Comment by jorgeorpinel** [comment]
p.s. I merged the Restyled version (correct formatting) to this PR so you need to `git pull` this branch (`tpi-blog`) on your end to make further changes. Thanks

**Comment by jendefig** [comment]
> p.s. I merged the Restyled version (correct formatting) to this PR so you need to `git pull` this branch (`tpi-blog`) on your end to make further changes. Thanks

Thanks @jorgeorpinel!

File: content/blog/2022-04-18-terraform-provider.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    sudo apt update
+    sudo apt install -y python3-pip
+    pip3 install -r requirements.txt
+    python3 src/train.py
+  END
+
+  storage {
+    workdir = "."
+    output  = "results"
+  }
```
Comment (casperdcl) [inline content/blog/2022-04-18-terraform-provider.md:81]:
```suggestion
  storage {
    workdir = "."
    output  = "results"
  }
  script = <<-END
    #!/bin/bash
    sudo apt-get update -q
    sudo apt-get install -yq python3-pip
    pip3 install -r requirements.txt
    python3 src/train.py
  END
```

File: content/blog/2022-04-18-terraform-provider.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+```hcl
+terraform {
+  required_providers {
+    iterative = {
+      source = "iterative/iterative"
+    }
+  }
+}
+
+provider "iterative" {}
```
Comment (casperdcl) [inline content/blog/2022-04-18-terraform-provider.md:63]:
```suggestion
  required_providers { iterative = { source = "iterative/iterative" } }
}
provider "iterative" {}
```

File: content/blog/2022-04-18-terraform-provider.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+you can use to configure:
+
+- Your cloud infrastructure
+- The steps to perform on the cloud resource, i.e. setting up the environment,
+  running the training pipeline, logging metrics, etc.
+- The data to be synced back once the training is complete (e.g. a file with
+  metrics, a model, plots)
+
+Here’s a “hello world” example of a `main.tf` Terraform configuration file using
+the `iterative_task` resource:
```
Comment (casperdcl) [inline content/blog/2022-04-18-terraform-provider.md:52]: should really link to example repo

**Review by casperdcl — CHANGES_REQUESTED** [review]

File: content/blog/2022-04-18-terraform-provider.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+## DevOps-Friendly
+
+Last, but not least, the Iterative Provider aims to bridge the gap between
+DevOps and Data Science teams. We build on top of Terraform, a tool universally
+familiar to DevOps teams, but extend it to suit ML needs.
+
+If you’d like to try the Iterative Provider in your project, check out the
+documentation on the provider’s page in the Terraform registry, and if you have
+any questions or suggestions, we welcome them in our
+[GitHub repo.](https://github.com/iterative/terraform-provider-iterative)
```
Comment (casperdcl) [inline content/blog/2022-04-18-terraform-provider.md:118]:
```suggestion
[GitHub repository](https://github.com/iterative/terraform-provider-iterative).
```

File: content/blog/2022-04-18-terraform-provider.md [hunk 6]
```diff
@@ -0,0 +1,124 @@
+---
+title: Machine Learning Workloads with Iterative Provider for Terraform
+date: 2022-04-18
+description: >
+  Today we introduce resource orchestration for your machine learning
+  projects  in conjunction with Terraform.
+descriptionLong: >
+  Now you can bring multi-cloud provisioning to your machine learning projects,
+  optimizing resources; costs, and workflows while aligning your Data
+  Science  and DevOps teams.
```
[the remark this answers is not in the document]
Comment (jendefig) [reply content/blog/2022-04-18-terraform-provider.md:10]: NOICE!!!! 🚀

File: content/blog/2022-04-18-terraform-provider.md (hunk 4 above)
Comment (jendefig) [reply content/blog/2022-04-18-terraform-provider.md:52]: Hmm.  Good point, is there one?  Does this code cover it?  Per our docs convo about superfluous repos, is it necessary?

File: content/blog/2022-04-27-terraform-provider.md [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+```hcl
+terraform {
+  required_providers {
+    iterative = {
+      source = "iterative/iterative"
+    }
+  }
+}
+
+provider "iterative" {}
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:63]:
```suggestion
  required_providers { iterative = { source = "iterative/iterative" } }
}
provider "iterative" {}
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    sudo apt update
+    sudo apt install -y python3-pip
+    pip3 install -r requirements.txt
+    python3 src/train.py
+  END
+
+  storage {
+    workdir = "."
+    output  = "results"
+  }
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:81]:
```suggestion
  storage {
    workdir = "."
    output  = "results"
  }
  script = <<-END
    #!/bin/bash
    sudo apt update
    sudo apt install -y python3-pip
    pip3 install --user -r requirements.txt
    python3 train.py
  END
```

**Review by casperdcl — CHANGES_REQUESTED** [review]

File: content/blog/2022-04-18-terraform-provider.md (hunk 4 above)
Comment (casperdcl) [reply content/blog/2022-04-18-terraform-provider.md]: it (used to) look like it was copy-pasted from an existing repo. Fine to leave a simpler version.

File: content/blog/2022-04-27-terraform-provider.md [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+```hcl
+terraform {
+  required_providers { iterative = { source = "iterative/iterative" } }
+}
+provider "iterative" {}
+
+resource "iterative_task" "example" {
+  cloud   = "aws"
+  region  = "us-east-2"
+  machine = "l+k80"
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:63]:
```suggestion
  cloud       = "aws"     # or any of: gcp, az, k8s
  machine     = "m"       # medium. Or any of: l, xl, m+k80, xl+v100, ...
  image       = "ubuntu"  # or "nvidia", ...
  region      = "us-east"
  disk_size   = 30        # GB
  spot        = 0         # auto-price. Or -1 to disable, or >0 for hourly USD limit
  timeout     = 60*60     # max 1h before forced termination
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+pipeline in a Terraform configuration file, you can bring that pipeline anywhere
+you want. You can use such a config for ad-hoc training at any stage of your
+prototyping process or use it as a job in your preferred CI/CD tool. It will
+also store your infrastructure configuration files in a version control system
+together with the rest of your project for easier control.
+
+## One Provider to Rule Them All
+
+Whether you prefer Amazon Web Services (AWS), Microsoft Azure, Google Cloud
+Platform (GCP), or Kubernetes, the Iterative Provider has you covered. You can
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:94]:
```suggestion
Platform (GCP), or Kubernetes (K8s), the Iterative Provider has you covered. You can
```

Comment (0x2b3bfa0) [reply content/blog/2022-04-27-terraform-provider.md:94]: > _One Provider to Rule Them All_

Clever for #devrel, “planeworthy” by and for #engineering. 💍

Comment (casperdcl) [reply content/blog/2022-04-27-terraform-provider.md:94]: One Provider to Rule Them All,
One Provider to `spot` them,
One ring to `apply` them all,
and in the darkness `destroy` them.

File: content/blog/2022-04-27-terraform-provider.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
+---
+title: Machine Learning Workloads with Iterative Provider for Terraform
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:2]:
```suggestion
title: Machine Learning Workloads with Terraform Provider Iterative
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+date: 2022-04-27
+description: >
+  Today we introduce painless resource orchestration for your machine learning
+  projects in conjunction with HashiCorp Terraform.
+descriptionLong: >
+  Catapult your machine learning projects into any coud and reduce cost with
+  spot instance auto-recovery, while aligning your Data Science  and DevOps
+  teams with a simple config file.
+picture: 2022-04-27/terraform-provider-iterative.png
+pictureComment: Machine Learning Workloads with Iterative Provider for Terraform
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:12]:
```suggestion
pictureComment: Machine Learning Workloads with Terraform Provider Iterative
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+machines, setting up environments and synchronizing data. For teams that want to
+leverage spot instances, the setup becomes even more complex - they need to make
+sure the training progress is not lost during spot instance recovery. This is
+time-consuming, and requires expertise in both DevOps and Machine Learning. On
+top of that, the costs of training in a cloud environment can be high due to not
+only the need for specific hardware, but also individual responsibility to shut
+down instances when training is complete.
+
+To address the specific needs of machine learning teams, we have built the
+Iterative Provider (TPI). TPI is an open-source tool extending the functionality
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:35]:
```suggestion
To address the specific needs of machine learning teams, we have built Terraform
Provider Iterative (TPI). TPI is an open-source tool extending the functionality
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+provider "iterative" {}
+
+resource "iterative_task" "example" {
+  cloud       = "aws"     # or any of: gcp, az, k8s
+  machine     = "m"       # medium. Or any of: l, xl, m+k80, xl+v100, ...
+  image       = "ubuntu"  # or "nvidia", ...
+  region      = "us-east"
+  disk_size   = 30        # GB
+  spot        = 0         # auto-price. Or -1 to disable, or >0 for hourly USD limit
+  timeout     = 60*60     # max 1h before forced termination
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:67]:
```suggestion
  region      = "us-west" # or us-west, eu-east, ...
  disk_size   = 30        # GB
  spot        = 0         # auto-price. Default -1 to disable, or >0 for hourly USD limit
  timeout     = 24*60*60  # max 24h before forced termination
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+}
+provider "iterative" {}
+
+resource "iterative_task" "example" {
+  cloud       = "aws"     # or any of: gcp, az, k8s
+  machine     = "m"       # medium. Or any of: l, xl, m+k80, xl+v100, ...
+  image       = "ubuntu"  # or "nvidia", ...
+  region      = "us-west" # or us-west, eu-east, ...
+  disk_size   = 30        # GB
+  spot        = 0         # auto-price. Default -1 to disable, or >0 for hourly USD limit
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:66]:
```suggestion
  spot        = 0         # auto-price. Default -1 to disable or >0 for hourly USD limit
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 16]
```diff
@@ -0,0 +1,122 @@
+---
+title: Machine Learning Workloads with Terraform Provider Iterative
+date: 2022-04-27
+description: >
+  Today we introduce painless resource orchestration for your machine learning
+  projects in conjunction with HashiCorp Terraform.
+descriptionLong: >
+  Catapult your machine learning projects into any coud and reduce cost with
+  spot instance auto-recovery, while aligning your Data Science  and DevOps
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:9]:
```suggestion
  spot instance auto-recovery, while aligning your Data Science and DevOps
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+  - Git
+  - Resource orchestation
+  - Spot Instance management
+---
+
+The requirements for Machine Learning (ML) infrastructure are becoming
+increasingly complex. Training large models often requires specialized hardware
+(GPUs, TPUs) which involves moving the whole training process onto cloud
+machines, setting up environments and synchronizing data. For teams that want to
+leverage spot instances, the setup becomes even more complex - they need to make
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:27]:
```suggestion
leverage spot instances, the setup becomes even more complex -- they need to make
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+sure the training progress is not lost during spot instance recovery. This is
+time-consuming, and requires expertise in both DevOps and Machine Learning. On
+top of that, the costs of training in a cloud environment can be high due to not
+only the need for specific hardware, but also individual responsibility to shut
+down instances when training is complete.
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:32]:
```suggestion
time-consuming, and requires expertise in both DevOps and Machine Learning.
Additionally, training in a cloud environment can incur high costs due to the
need for expensive hardware, as well as users forgetting to shutdown instances
when training is complete.
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+
+To address the specific needs of machine learning teams, we have built Terraform
+Provider Iterative (TPI). TPI is an open-source tool extending the functionality
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:35]:
```suggestion
To address the specific needs of machine learning teams, we have built
[Terraform Provider Iterative (TPI)](https://github.com/iterative/terraform-provider-iterative).
TPI is an open-source tool extending the functionality
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+
+Once the training is complete, the Iterative Provider terminates the resource,
+so users don't have to worry about spiraling costs from unused machines.
+
+## Configure Once, Bring Everywhere
+
+Once you configure infrastructure and a script that executes your training
+pipeline in a Terraform configuration file, you can bring that pipeline anywhere
+you want. You can use such a config for ad-hoc training at any stage of your
+prototyping process or use it as a job in your preferred CI/CD tool. It will
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:91]:
```suggestion
prototyping process or use it as a job in your preferred CI/CD tool. You can
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+
+## One Provider to Rule Them All
+
+Whether you prefer Amazon Web Services (AWS), Microsoft Azure, Google Cloud
+Platform (GCP), or Kubernetes (K8s), the Iterative Provider has you covered. You
+can configure compute resources from these with a single provider, using
+[common machine types](https://registry.terraform.io/providers/iterative/iterative/latest/docs/resources/task#machine-type).
+These are going to be roughly similar between different cloud vendors. Not only
+does this significantly simplify infrastructure configuration, it also makes it
+easy to migrate from one cloud to another with only minor adjustments.
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:103]:
```suggestion
can configure compute resources from these with a unified API, using
[common machine types](https://registry.terraform.io/providers/iterative/iterative/latest/docs/resources/task#machine-type)
that are the same across all cloud vendors. This significantly simplifies
infrastructure configuration and makes it easy to migrate from one cloud to another
by changing just one line of code.
```

File: content/blog/2022-04-27-terraform-provider.md [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+
+## Costs Optimization
+
+The Iterative Provider helps with cloud compute cost optimization in two major
+ways. First, upon completion of your script, the instance is automatically
+terminated. This helps to avoid accumulating costs due to abandoned resources.
+Second, you can leverage the cost-saving power of spot instances to train your
+models without losing any progress!
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:111]:
```suggestion
models without losing any progress! TPI recovers the working directory and
respawns interrupted/preempted instances for you.
```

**Review by casperdcl — CHANGES_REQUESTED** [review]
copyedits

**Review by casperdcl — APPROVED** [review]

File: content/blog/2022-04-27-terraform-provider.md [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+picture: 2022-04-27/terraform-provider-iterative.png
+pictureComment: Machine Learning Workloads with Terraform Provider Iterative
+author: maria_khalusova
+commentsUrl: https://discuss.dvc.org/t/terraform-iterative-provider-release-blog-post/1171
+tags:
+  - MLOps
+  - Terraform
+  - Git
+  - Resource orchestation
+  - Spot Instance management
```
Comment (casperdcl) [inline content/blog/2022-04-27-terraform-provider.md:20]:
```suggestion
  - MLOps
  - Terraform
  - Git
  - Cloud orchestration
  - Spot instance management
```

**Comment by casperdcl** [comment]
live @ https://dvc.org/blog/terraform-provider

**Pull request merged**